### PR TITLE
feat(api): fixed patch upload by id endpoint

### LINKED ISSUE #3041: feat/API: missing patchUploadbyId method in API (corresponding to organize -> uploads -> edit properties)
in REST API V2 https://github.com/fossology/fossology/blob/master/src/www/ui/api/documentation/openapiv2.yaml the following methods are currently implemented within the "Organize" category/tag:

- createFolder
- deleteFolderById
- moveFolderById
- patchFolderById -> change name and description of a folder, not of an upload
- createGroup

- deleteUploadById
- moveUploadById
- updateUploadById -> change upload status and assignee, not name and description

A patchUploadbyId method is missing, that should allow to change name and description of an upload via API

### PR DESCRIPTION
**Pull request by harshitg927**
## Description

This PR modifies the updateUpload method in UploadController to properly handle all parameters from the request body instead of query parameters. This allows users to update upload status, assignee, name, and description via a single JSON payload, making the API more RESTful and consistent with the Swagger documentation.

### Changes

- Modified updateUpload method to read all parameters from the request body instead of query parameters
- Simplified the handling of comments for status changes
- Updated test cases to use proper JSON request format
- Fixed the API documentation to match the implementation
- Ensured proper status code (202) is returned with success message

## How to test

1. Test the API endpoint by sending a PATCH request to repo/api/v2/uploads/{uploadId} with a JSON body containing any of the following fields:
{
  "status": "Closed",
  "assignee": 3,
  "comment": "Upload reviewed and closed",
  "name": "new_upload_name.zip",
  "uploadDescription": "Updated description"
}
2. Verify that each field properly updates in the UI (http://localhost/repo/?mod=browse)

Screenshot from 2025-04-30 14-46-08

Fixes: [issue#3041](https://github.com/fossology/fossology/issues/3041)

### CODE REVIEW

**Comment by its-sushant** [comment]
@harshitg927 It looks like there are a lot of formatting changes in this PR, especially around tab/indentation length. These changes make the diff harder to review and aren't related to the actual functionality. Could you revert these formatting changes and keep the diff focused on the functional updates?

**Comment by harshitg927** [comment]
Hello @its-sushant , I made the indentation changes because of the checks failing for codesniffer, if I revert back the changes the checks may fail. 
Should I still revert the formatting changes?

**Comment by its-sushant** [comment]
> Hello @its-sushant , I made the indentation changes because of the checks failing for codesniffer, if I revert back the changes the checks may fail. Should I still revert the formatting changes?

Yes @harshitg927 You can revert those changes for now!

**Comment by harshitg927** [comment]
@its-sushant I have made the necessary formatting changes, PTAL.

File: src/www/ui/api/Controllers/UploadController.php [hunk 1]
```diff
@@ -679,30 +678,25 @@ public function updateUpload($request, $response, $args)
...
-      $assignee = filter_var($query[self::FILTER_ASSIGNEE], FILTER_VALIDATE_INT);
+    if (array_key_exists(self::FILTER_ASSIGNEE, $bodyContent)) {
+      $assignee = filter_var($bodyContent[self::FILTER_ASSIGNEE], FILTER_VALIDATE_INT);
       $userList = $userDao->getUserChoices($groupId);
-      if (!array_key_exists($assignee, $userList)) {
-        throw new HttpNotFoundException(
-          "New assignee does not have permission on upload.");
+    if (!array_key_exists($assignee, $userList)) {
+      throw new HttpNotFoundException(
+        "New assignee does not have permission on upload.");
```
Comment (its-sushant) [inline src/www/ui/api/Controllers/UploadController.php:686]: Same here!

File: src/www/ui/api/Controllers/UploadController.php [hunk 2]
```diff
@@ -679,30 +678,25 @@ public function updateUpload($request, $response, $args)
...
-        if ($isJsonRequest && array_key_exists("comment", $bodyContent)) {
-          $comment = $bodyContent["comment"];
-        } else {
-          $comment = $bodyContent;
-        }
-      }
+      array_key_exists(self::FILTER_STATUS, $bodyContent) &&
+      in_array(strtolower($bodyContent[self::FILTER_STATUS]), self::VALID_STATUS)
+      ) {
+      $newStatus = strtolower($bodyContent[self::FILTER_STATUS]);
```
Comment (its-sushant) [inline src/www/ui/api/Controllers/UploadController.php:699]:
```suggestion
      array_key_exists(self::FILTER_STATUS, $bodyContent) &&
      in_array(strtolower($bodyContent[self::FILTER_STATUS]), self::VALID_STATUS)
    ) {
      $newStatus = strtolower($bodyContent[self::FILTER_STATUS]);
```

File: src/www/ui/api/Controllers/UploadController.php [hunk 3]
```diff
@@ -713,24 +707,23 @@ public function updateUpload($request, $response, $args)
       } else {
         $status = UploadStatus::OPEN;
       }
-      $uploadBrowseProxy->setStatusAndComment($id, $status, $comment);
+        $uploadBrowseProxy->setStatusAndComment($id, $status, $comment);
```
Comment (its-sushant) [inline src/www/ui/api/Controllers/UploadController.php:710]: Same.

File: src/www/ui/api/Controllers/UploadController.php [hunk 4]
```diff
@@ -713,24 +707,23 @@ public function updateUpload($request, $response, $args)
...
       $newName = trim($bodyContent[self::FILTER_NAME]);
     }
-    // Handle update of description
+
     if (
-      $isJsonRequest &&
       array_key_exists("uploadDescription", $bodyContent) &&
       strlen(trim($bodyContent["uploadDescription"])) > 0
-    ) {
+      ) {
```
Comment (its-sushant) [inline src/www/ui/api/Controllers/UploadController.php:723]:
```suggestion
    ) {
```

File: src/www/ui_tests/api/Controllers/UploadControllerTest.php [hunk 5]
```diff
@@ -1183,50 +1183,69 @@ private function testGetUploadLicensesPendingScan($version = ApiVersion::V2)
...
-   * -# Check if response status is 202
-   */
-  public function testUpdateUpload()
-  {
+ * @test
+ * -# Test for UploadController::updateUpload()
+ * -# Check if response status is 202
+ */
+public function testUpdateUpload()
+{
```
Comment (its-sushant) [inline src/www/ui_tests/api/Controllers/UploadControllerTest.php:1191]: Please fix the indentation. There are lot of indentation changes here which is not required in this function. Please take a look!

**Review by its-sushant — CHANGES_REQUESTED** [review]
@harshitg927 Please take a look into the suggested changes!

**Comment by its-sushant** [comment]
Hello @harshitg927. Any update on this.

**Comment by harshitg927** [comment]
@its-sushant Hello, sorry for the delay, there were some issues with my computer, I will fix the linting issues and push the code ASAP.

File: src/www/ui/api/Controllers/UploadController.php [hunk 6]
```diff
@@ -715,22 +709,21 @@ public function updateUpload($request, $response, $args)
       }
       $uploadBrowseProxy->setStatusAndComment($id, $status, $comment);
     }
-    // Handle update of name
```
Comment (Kaushl2208) [inline src/www/ui/api/Controllers/UploadController.php:718]: Is there a reason behind removing the comment:  // Handle update of name?

File: src/www/ui/api/Controllers/UploadController.php [hunk 7]
```diff
@@ -679,30 +678,25 @@ public function updateUpload($request, $response, $args)
...
-      $newStatus = strtolower($query[self::FILTER_STATUS]);
-      $comment = '';
-      if (in_array($newStatus, ["closed", "rejected"])) {
-        if ($isJsonRequest && array_key_exists("comment", $bodyContent)) {
-          $comment = $bodyContent["comment"];
-        } else {
-          $comment = $bodyContent;
-        }
-      }
+      $newStatus = strtolower($bodyContent[self::FILTER_STATUS]);
```
Comment (Kaushl2208) [inline src/www/ui/api/Controllers/UploadController.php:702]: Logic for validating a comment for specific status: Closed or Rejected was mandatory. We should keep that, There has to be a specific comment provided for these two specific Status. As to why rejected or closed

File: src/www/ui/api/Controllers/UploadController.php [hunk 8]
```diff
@@ -679,30 +678,25 @@ public function updateUpload($request, $response, $args)
...
+      $assignee = filter_var($bodyContent[self::FILTER_ASSIGNEE], FILTER_VALIDATE_INT);
       $userList = $userDao->getUserChoices($groupId);
       if (!array_key_exists($assignee, $userList)) {
         throw new HttpNotFoundException(
           "New assignee does not have permission on upload.");
       }
       $uploadBrowseProxy->updateTable("assignee", $id, $assignee);
     }
-    // Handle new status
+
```
Comment (Kaushl2208) [inline src/www/ui/api/Controllers/UploadController.php:691]: Also type checking if `comment` is actual string would make more sense.

**Review by Kaushl2208 — CHANGES_REQUESTED** [review]
Few changes that you should look into. Logic change should follow the mandatory requirements from before.

File: src/www/ui/api/Controllers/UploadController.php [hunk 9]
```diff
@@ -715,22 +709,21 @@ public function updateUpload($request, $response, $args)
       }
       $uploadBrowseProxy->setStatusAndComment($id, $status, $comment);
```
Comment (Kaushl2208) [inline src/www/ui/api/Controllers/UploadController.php:733]: What if the user didnt provide the comment in request body? There is no initialization of the comment?

**Comment by Kaushl2208** [comment]
@harshitg927 , Any updates on this?

File: src/www/ui/api/Controllers/UploadController.php (hunk 6 above)
Comment (harshitg927) [reply src/www/ui/api/Controllers/UploadController.php:718]: I have included the comments which were previously there along with some comments for the work done in this pull request.

File: src/www/ui/api/Controllers/UploadController.php (hunk 7 above)
Comment (harshitg927) [reply src/www/ui/api/Controllers/UploadController.php:702]: Done

File: src/www/ui/api/Controllers/UploadController.php (hunk 8 above)
Comment (harshitg927) [reply src/www/ui/api/Controllers/UploadController.php:691]: Done

File: src/www/ui/api/Controllers/UploadController.php (hunk 9 above)
Comment (harshitg927) [reply src/www/ui/api/Controllers/UploadController.php:733]: Done, I have initialized the comment field with empty string.

**Comment by harshitg927** [comment]
@Kaushl2208 , I have made the changes requested, PTAL.

File: src/www/ui/api/documentation/openapiv2.yaml [hunk 10]
```diff
@@ -549,59 +549,60 @@ paths:
...
+                  description: New status of the upload
+                  example: Closed
+                assignee:
+                  type: integer
+                  description: ID of new assignee for the upload
+                  example: 3
                 comment:
                   type: string
-                  description: The comment for new status
+                  description: Optional comment for the status change
```
Comment (Kaushl2208) [inline src/www/ui/api/documentation/openapiv2.yaml]: Umm something Like this makes more sense right?

> Required when status is 'Closed' or 'Rejected'.

**Review by Kaushl2208 — CHANGES_REQUESTED** [review]
Hey @harshitg927 , Most of the things look good. There are few edge cases that I was trying:

image

- There should be an exception handling: Invalid Status Type right???

As for Controller Test cases:

- Missing or empty comment for Closed/Rejected
- Comment not a string
- Invalid status
- Name/description missing or empty

These edge cases will make good reliable coverage.

**Comment by shaheemazmalmmd** [comment]
@harshitg927 : Please look into the review comments my @Kaushl2208 .

**Comment by harshitg927** [comment]
@Kaushl2208 I have made the required change, PTAL

**Comment by harshitg927** [comment]
@Kaushl2208 Are there any updates on this pull request?

File: src/www/ui/api/documentation/openapiv2.yaml [hunk 11]
```diff
@@ -549,59 +549,60 @@ paths:
...
+                  description: New status of the upload
+                  example: Closed
+                assignee:
+                  type: integer
+                  description: ID of new assignee for the upload
+                  example: 3
                 comment:
                   type: string
-                  description: The comment for new status
+                  description: Required when status is 'Closed' or 'Rejected'.
```
Comment (Kaushl2208) [inline src/www/ui/api/documentation/openapiv2.yaml:583]: `Comment for new status. Required when status is 'Closed' or 'Rejected'`

**Review by Kaushl2208 — CHANGES_REQUESTED** [review]
Code looks good, Needs test!!
Also please rebase with the latest master.

File: src/www/ui/api/Controllers/UploadController.php [hunk 12]
```diff
@@ -680,29 +679,47 @@ public function updateUpload($request, $response, $args)
     }
 
     // Handle assignee info
-    if (array_key_exists(self::FILTER_ASSIGNEE, $query)) {
-      $assignee = filter_var($query[self::FILTER_ASSIGNEE], FILTER_VALIDATE_INT);
+    if (array_key_exists(self::FILTER_ASSIGNEE, $bodyContent)) {
```
Comment (Kaushl2208) [inline src/www/ui/api/Controllers/UploadController.php:682]: How are we getting this `$bodyContent` variable?

File: src/www/ui/api/Controllers/UploadController.php [hunk 13]
```diff
@@ -648,7 +648,6 @@ public function getUploadCopyrights($request, $response, $args)
   public function updateUpload($request, $response, $args)
   {
     $id = intval($args['id']);
-    $query = $request->getQueryParams();
```
Comment (Kaushl2208) [inline src/www/ui/api/Controllers/UploadController.php:651]: You have removed the `$query` but you will also need to read body content?

File: src/www/ui/api/Controllers/UploadController.php (hunk 12 above)
Comment (Kaushl2208) [reply src/www/ui/api/Controllers/UploadController.php:682]: if we have removed `text/plain` and only rely on `application/json` please look into the check how we used to handle it before.

**Review by Kaushl2208 — COMMENTED** [review]
Hey @harshitg927 , i hope you tested the changes again after pushing the fix. I am leaving some of the changes here. Please fix and test as well

File: src/www/ui/api/Controllers/UploadController.php (hunk 12 above)
Comment (harshitg927) [reply src/www/ui/api/Controllers/UploadController.php:682]: @Kaushl2208 we are getting the $bodyContent variable from a if block above which is retrieving the body content using the $this->getParsedBody($request); method.
this is around line 674 in the upload controller itself.

File: src/www/ui/api/Controllers/UploadController.php (hunk 13 above)
Comment (harshitg927) [reply src/www/ui/api/Controllers/UploadController.php:651]: @Kaushl2208 , the body content is already being read at around line 674 in this if statement
if ($isJsonRequest) {
      $bodyContent = $this->getParsedBody($request);
}
please take a look.

File: src/www/ui/api/Controllers/UploadController.php (hunk 12 above)
Comment (harshitg927) [reply src/www/ui/api/Controllers/UploadController.php:682]: I will take a look into how we handle only relying on application/json format in other parts of code
